torch_object_detection_segmenter.py

```diff
@@ -1,31 +1,191 @@
+                h_beg = np.random.randint(0, img_h - target_h + 1)
+            elif how == 'precise':
+                assert (w_beg is not None and h_beg is not None)
+                assert (0 <= w_beg <= (img_w - target_w)), f'left must be within [0, {img_w - target_w}]: {w_beg}'
+                assert (0 <= h_beg <= (img_h - target_h)), f'top must be within [0, {img_h - target_h}]: {h_beg}'
+            else:
+                raise ValueError(f'unknown input how: {how}')
+            if not isinstance(w_beg, int):
+                raise ValueError(f'left must be int number between 0 and {img_w}: {left}')
+            if not isinstance(h_beg, int):
+                raise ValueError(f'top must be int number between 0 and {img_h}: {top}')
+            w_end = w_beg + target_w
+            h_end = h_beg + target_h
+            img = img.crop((w_beg, h_beg, w_end, h_end))
+            return img, h_beg, w_beg
+
         if not docs:
             return
-        for doc in docs:
-            assert doc.blob.shape[0] == 256
-            doc.embedding = np.random.rand(self._dim)
\ No newline at end of file
+
+        batch = docs.get_attributes('blob') # (2, 681, 1264, 3) with imgs/cars.jpg
+        # "Ensure the color channel axis is the default axis." i.e. c comes first
+        # e.g. (h,w,c) -> (c,h,w) / (b,h,w,c) -> (b,c,h,w)
+        batch = _move_channel_axis(batch, self.channel_axis, self._default_channel_axis + 1) # take batching into account
+
+        batched_predictions = self._predict(batch)
+
+
+        for i, (image, predictions) in enumerate(zip(batch, batched_predictions)):
+            bboxes = predictions['boxes'].detach()
+            scores = predictions['scores'].detach()
+            labels = predictions['labels']
+            if self.on_gpu:
+                bboxes = bboxes.cpu()
+                scores = scores.cpu()
+                labels = labels.cpu()
+            img = _load_image(image * 255, self._default_channel_axis)
+
+            for bbox, score, label in zip(bboxes.numpy(), scores.numpy(), labels.numpy()):
+                if score >= self.confidence_threshold:
+                    x0, y0, x1, y1 = bbox
+                    # note that tensors are [H, W] while PIL Images are [W, H]
+                    top, left = int(y0), int(x0)
+                    # target size must be (h, w)
+                    target_size = (int(y1) - int(y0), int(x1) - int(x0))
+                    # at this point, raw_img has the channel axis at the default tensor one
+                    _img, top, left = _crop_image(img, target_size=target_size, top=top, left=left, how='precise')
+                    _img = _move_channel_axis(np.asarray(_img).astype('float32'), -1, self.channel_axis)
+                    label_name = self.label_name_map[label]
+                    logging.debug(
+                        f'detected {label_name} with confidence {score} at position {(top, left)} and size {target_size}')
```

Comment:
u need to instantiate a JinaLogger in init

Reply:
I don't think the JinaLogger from helper is available in 2.0 any more

Reply:
it is

Reply:
added